### PR TITLE
[SPARK-23772][SQL] Provide an option to ignore column of all null values or empty array during JSON schema inference

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added a new JSON option `dropFieldIfAllNull ` to ignore column of all null values or empty array/struct during JSON schema inference. 

## How was this patch tested?
Added tests in `JsonSuite`.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala]: I just added this weird type for this prototype, so we need to consider more here. I don't have better idea now, so I'd like to have any suggestion about this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: * I was expecting a configuration for the JSON data source instead of a Spark SQL conf. Say, how to implement this feature if the data source is outside Spark?
* Since Spark SQL cannot control the behavior of external data source implementations, the flag won't work unless external data sources recognize it. This is another reason to put it under JSON data source.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `ignoreNullFields` is not precise. For example

```json
{"a": 1}
{"a": null}
```

The `null` value is not ignored. I would suggest `dropFieldIfAllNull`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (mengxr) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: We need more test coverage. I have a similar internal implementation that tests the following cases (ignore the actual test, just look at the example records):

```scala
  test("null") {
    assert(removeNullRecursively("null") === "null")
  }

  test("empty string") {
    assert(removeNullRecursively("\"\"") === "\"\"")
  }

  test("empty object") {
    assert(removeNullRecursively("{}") === "null")
  }

  test("object with all null values") {
    val json = """{"a":null,"b":null, "c":null}"""
    assert(removeNullRecursively(json) === "null")
  }

  test("object with some null fields") {
    val json = """{"a":null,"b":"c","d":null,"e":"f"}"""
    val expected = """{"b":"c","e":"f"}"""
    assert(removeNullRecursively(json) === expected)
  }

  test("object with some nested null values") {
    val json = """{"a":{},"b":{"c":null},"d":{"c":"e"},"f":{"c":null,"g":"h"}}"""
    val expected = """{"d":{"c":"e"},"f":{"g":"h"}}"""
    assert(removeNullRecursively(json) === expected)
  }

  test("array with all null elements") {
    val json = """[null,null,{},{"a":null}]"""
    val expected = "null"
    assert(removeNullRecursively(json) === expected)
  }

  test("array with some null elements") {
    // TODO: is it an issue if we covert empty object to null in an array?
    val json = """[null,"a",null,{},"b"]"""
    val expected = """[null,"a",null,null,"b"]"""
    assert(removeNullRecursively(json) === expected)
  }
```

**Review by mengxr — CHANGES_REQUESTED** [review]
Sorry for late review! I just saw it ... My main feedback is that this feature shouldn't require adding new types to Spark SQL, nor adding new Spark SQL conf. It should live in the JSON data source. I also recommended a way to implement it in inline comments.

**Comment by maropu** [comment]
Thanks for the review! ok, plz give me more time to reconsider the design based the comments.

**Comment by maropu** [comment]
oh, sorry, I'll update in a few days.

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok, I dropped the conf in `SQLConf`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok, I used `dropFieldIfAllNull`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala]: In the first attempt, I used the new type instead of `NullType` because some `Sink`s (`FileStreamSink`) could not handle `NullType`;
```
// parquet
java.lang.RuntimeException: Unsupported data type NullType.
        at scala.sys.package$.error(package.scala:27)
        at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport.org$apache$spark$sql$execution$datasources$parquet$ParquetWriteSupport$$makeWriter(ParquetWriteSupport.scala:206)
        at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$init$2.apply(ParquetWriteSupport.scala:93)
        at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$init$2.apply(ParquetWriteSupport.scala:93)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
// orc
java.lang.IllegalArgumentException: Can't parse category at 'struct<c0:bigint,c1:null^,c2:array<null>>'
        at org.apache.orc.TypeDescription.parseCategory(TypeDescription.java:223)
        at org.apache.orc.TypeDescription.parseType(TypeDescription.java:332)
        at org.apache.orc.TypeDescription.parseStruct(TypeDescription.java:327)
        at org.apache.orc.TypeDescription.parseType(TypeDescription.java:385)
        at org.apache.orc.TypeDescription.fromString(TypeDescription.java:406)

// csv
java.lang.UnsupportedOperationException: CSV data source does not support null data type.
        at org.apache.spark.sql.execution.datasources.csv.CSVUtils$.org$apache$spark$sql$execution$datasources$csv$CSVUtils$$verifyType$1(CSVUtils.scala:130)
        at org.apache.spark.sql.execution.datasources.csv.CSVUtils$$anonfun$verifySchema$1.apply(CSVUtils.scala:134)
        at org.apache.spark.sql.execution.datasources.csv.CSVUtils$$anonfun$verifySchema$1.apply(CSVUtils.scala:134)
        at scala.collection.Iterator$class.foreach(Iterator.scala:893)

```
So, in the previous fix, I tried to add `PlaceholderType` inherited from `StringType` and this type could be correctly handled in all the `Sink`, but too tricky. 

In the suggested, `NullType, ArrayType(NullType), etc should be dropped` means that we need to handle an inferred schema as follows? e.g.,
```
Inferred schema: "StructType<IntegerType, NullType, ArrayType(NullType)>" -> Schema used in FileStreamSource: "StructType<IntegerType>"
```
Is this right?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala]: I think `removeNullRecursively` seems to be a little time-consuming (json parsing twice) and why can't we do the same thing directly in `JacksonParser`? e.g.,
https://github.com/apache/spark/pull/20929/files#diff-635e02b2d1ce4ad1675b0350ccac0c10R334

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/TypePlaceholder.scala]: In the latest fix, I used `NullType` for unresolved types and `FileStreamSource` would resolve the types when we see non-null values. As suggested in `SPARK-12436`, I think we might need to fallback into `StringType` just before `Sink`s, e.g., https://github.com/apache/spark/blob/76ecd095024a658bf68e5db658e4416565b30c17/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala#L492

**Comment by mengxr** [comment]
@maropu Thanks for the update! I saw some test failures, no sure if they are relevant. Looking at the changes, I feel we don't need to handle the streaming scenarios in the same PR. We just add a flag to JSON data source and leave it off by default, so it won't break any existing streaming use cases. In this PR, we should focus on properly identifying "null" types, which include NullType, ArrayType(NullType), ArrayType(ArrayType(NullType)), etc.

Could you make another update? Thanks!

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
ok, in this pr, I'll focus on adding a new flag to do so. just a sec for the update. Thanks!

**Comment by maropu** [comment]
@mengxr Could you check? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: How about [DataStreamReader](https://github.com/maropu/spark/blob/SPARK-23772/sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala#L277)? I guess the description should be added to it too.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Do you need the `overwrite` mode?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: It seems the `DefaultEquality` is used here which applies `==`. Are there any reasons for `===` instead of just `==`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Could you add a test when `dropFieldIfAllNull` is set to `true` but not all values in a column are `null`s

**Review by MaxGekk — COMMENTED** [review]
Could you add description for the option to readwriter.py and streaming.py

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Python side too.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: No, there's no explicit preference between them since the preferences are diverted even in committers. It's fine to use one of them.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: ok

**Comment by mengxr** [comment]
@maropu Thanks for updating this PR! It would be easier to maintain the logic in one place. I think it should be feasible to do everything inside `canonicalizeType` without modifying `JsonParser` or other methods in `JsonInferSchema`. The following code outlines my logic, though I didn't test it ...:

~~~scala
 /**
   * Canonicalize data types and remove StructTypes with no fields.
   * @return Some(canonicalizedType) or None if nothing left.
   */
  private def canonicalizeType(tpe: DataType, options: JSONOptions): Option[DataType] = tpe match {
    case at @ ArrayType(elementType, _) =>
      canonicalizeType(elementType, options).map(t => at.copy(elementType = t))

    case StructType(fields) =>
      val canonicalizedFields = fields.flatMap { f =>
        canonicalizeType(f.dataType, options).map(t => f.copy(dataType = t))
      }
      // per SPARK-8093: empty structs should be deleted
      if (canonicalizedFields.isEmpty) {
        None
      } else {
        StructType(canonicalizedFields)
      }

    case NullType => 
      if (options.dropFieldIfAllNull) {
        None
      else {
        Some(StringType)
      }
    }

    case other => Some(other)
  }  
~~~

In the test, we should also include scenarios with nested "null" fields like `[[], null, [[]]]`.

**Comment by maropu** [comment]
yea, thanks for the comments! I'll try to fix based on the comments.

**Comment by maropu** [comment]
sorry, I'll update tongiht

**Comment by maropu** [comment]
@mengxr ok, could you check?

**Comment by mengxr** [comment]
@maropu I sent you a PR at https://github.com/maropu/spark/pull/1. Could you take a look and see if there are things I missed. I updated the tests, so the null fields are dropped in the expected schema.

**Comment by maropu** [comment]
ok, I'll check

**Comment by maropu** [comment]
I checked the json test passed in my local. So, I merged the fix into this. Sorry to bother you, but could you check again?

**Comment by maropu** [comment]
@HyukjinKwon I think this is your area, so could you double-check this? Thanks!

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by mengxr** [comment]
LGTM pending Jenkins

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
